text/2021-04-23-builtin-gc-worker.md

```diff
@@ -0,0 +1,131 @@
+
+1. `GC Barrier`
+
+    PD guarantees that the GC process will not affect the data with a timestamp later than any `GC Barrier`. TiKV users (e.g. TiDB / CDC / Client) can put GC Barriers into PD so as to make sure the data they are using will not be GC.
+
+2. `Transaction Safepoint`
+
+    TiKV and TiFlash guarantee that no read or write will happen on the data earlier than `Transaction Safepoint`. PD calculates `Transaction Safepoint Proposal` by `min(GC Barriers) - 1` and proposes it to TiKV and TiFlash via the `StoreHeartbeatResponse`. Therefore, TiKV and TiFlash will reject all requests with timestamp earlier than the `Transaction Safepoint`. In the next following `StoreHeartbeatRequest` TiKV and TiFlash will report their latest `Transaction Safepoint Status` to PD.
+
+3. `GC Safepoint`
+
+    PD calculates `GC Safepoint` by `min(Transaction Safepoint Status) - 1`, resolves locks earlier than `GC Safepoint`, then synchronizes it to TiKV and TiFlash via `StoreHeartbeatResponse`. And finally, TiKV and TiFlash are safe to delete MVCC data earlier than `GC Safepoint`.
+
+![GC Worker](../media/gc-worker.png)
+
+According to the design above, several fields will be added to `pdpb.proto`:
+
+```diff
+  // `pdpb.proto
+
+  message StoreHeartbeatRequest {
+      RequestHeader header = 1;
+      StoreStats stats = 2;
+  }
+
+  message StoreStats {
+  …
++     uint64 txn_safepoint = 15;
++     uint64 gc_safepoint = 16;
+  }
+
+  message PutStoreResponse {
+      ResponseHeader header = 1;
+      replication_modepb.ReplicationStatus replication_status = 2;
++     uint64 txn_safepoint = 3;
++     uint64 gc_safepoint = 4;
+  }
+
+  message StoreHeartbeatResponse {
+     ResponseHeader header = 1;
+     replication_modepb.ReplicationStatus replication_status = 2;
+     string cluster_version = 3;
++    uint64 txn_safepoint = 4;
++    uint64 gc_safepoint = 5;
+  }
+```
+
+Aslo, RPC `SetGCBarrier` and `GetGCStatus` will be added:
+
+```diff
+  // `pdpb.proto
+
+  service PD {
+  ...
++    rpc SetGCBarrier(SetGCBarrierRequest) returns (SetGCBarrierResponse) {}
++    rpc GetGCStatus(GetGCStatusRequest) returns (GetGCStatusResponse) {}
+  }
+
++ message GCBarrier {
++    bytes id = 1;
++    // Set to 0 to delete existing barrier
++    uint64 ttl = 2;
++    uint64 timestamp = 3;
++ }
+
++ message SetGCBarrierRequest {
++    RequestHeader header = 1;
++    GCBarrier gc_barrier = 2;
++ }
+
++ message SetGCBarrierResponse {
++    ResponseHeader header = 1;
++ }
+
++ message GetGCStatusRequest {
++    RequestHeader header = 1;
++ }
+
++ message GetGCStatusResponse {
++    ResponseHeader header = 1;
++    repeated GCBarriers gc_barriers = 2;
++    uint64 txn_safepoint = 3;
++    uint64 gc_safepoint = 4;
++ }
+```
+
+### Backward compatibility
+
+#### Rolling update
+
+During the rolling update, the GC worker will not function and the new TiDB will not try to be elected as a GC leader. Once the rolling update is completed, the GC worker in PD will start to work.
```

Comment:
How do you check if the rolling update is complete?

Reply:
I think it's similar to async commit which has to be turned on manually. One problem is that  async commit switch is written in TiDB global variable which is not viable w/o TiDB. Maybe we have to add a pdpb RPC for that or to use etcd.

Reply:
I'm afraid it would be over-complicated, and currently we didn't support automatically  enabling async commit on rolling  update due to the difficulty. Can you try to make them both able to work without breaking any consistency?

Reply:
Yeah, it is difficult, that's why I want to turned it on manually after rolling update. Once pd has turned builtin gc worker on, it will then reject UpdateGCSafepoint request.

Reply:
Emmm... personally I think it would not be a good idea if the user need to update the configuration manually, since GC should be as transparent as possible.

Reply:
That's similar to the async commit switch which is a must-do. In detail, it can be a specified key in etcd.